File: src/da4ml/cmvm/core/indexers.py
import numpy as np
from numba import jit

from ..types import DAState, QInterval, minimal_kif


@jit
def idx_mc(state: DAState):
    """Choose the pair with highest frequency."""
    freqs = list(state.freq_stat.values())
    max_freq = max(freqs)
    pair_idx = freqs.index(max_freq)
    return pair_idx


@jit
def idx_mc_dc(state: DAState, absolute: bool = False):
    """Choose the pair with highest frequency with latency penalty.
    If absolute is True, return -1 if any latency overhead may present."""
    freqs = list(state.freq_stat.values())
    if max(freqs) <= 2:
        return -1
    factor = max(freqs) + 1
    ops = state.ops
    lat_penalty = [abs(ops[pair.id1].latency - ops[pair.id0].latency) * factor for pair in state.freq_stat.keys()]
    score = [freq - lat_penalty[i] for i, freq in enumerate(freqs)]
    max_score = max(score)
    if absolute and max_score < 0:
        return -1
    pair_idx = score.index(max_score)
    return pair_idx


@jit
def overlap_and_total(qint0: QInterval, qint1: QInterval):
    """Calculate the overlap and total number of bits for two QIntervals, when represented in fixed-point format."""
    k0, i0, f0 = minimal_kif(qint0)
    k1, i1, f1 = minimal_kif(qint1)
    n_overlap = max(min(k0, k1) + min(i0, i1) + min(f0, f1), 0)
    n_total = max(k0, k1) + max(i0, i1) + max(f0, f1)
    return n_overlap, n_total


@jit
def idx_wmc(state: DAState):
    """Choose the pair with the highest weighted most common subexpression (WMC) score."""
    freqs = list(state.freq_stat.values())
    keys = list(state.freq_stat.keys())
    score = np.empty(len(freqs), dtype=np.float32)
    for i, (k, v) in enumerate(zip(keys, freqs)):
        id0, id1 = k.id0, k.id1
        qint0, qint1 = state.ops[id0].qint, state.ops[id1].qint
        n_overlap, n_total = overlap_and_total(qint0, qint1)
        score[i] = v * (2 * n_overlap - n_total)
    max_score = np.max(score)
    if max_score < 0:
        return -1
    return np.argmax(score)


@jit
def idx_wmc_dc(state: DAState, absolute: bool = False):
    """Choose the pair with the highest weighted most common subexpression (WMC) score with latency and cost penalty.
    When absolute is True, return -1 if any latency overhead may present."""
    freqs = list(state.freq_stat.values())
    if max(freqs) == 2:
        return -1
    keys = list(state.freq_stat.keys())
    score = np.empty(len(freqs), dtype=np.float32)
    for i, (k, v) in enumerate(zip(keys, freqs)):
        id0, id1 = k.id0, k.id1
        qint0, qint1 = state.ops[id0].qint, state.ops[id1].qint
        lat0, lat1 = state.ops[id0].latency, state.ops[id1].latency
        n_overlap, n_total = overlap_and_total(qint0, qint1)
        score[i] = v * (2 * n_overlap - n_total) - 256 * abs(lat0 - lat1)
    if absolute and np.max(score) < 0:
        return -1
    return np.argmax(score)

File: src/da4ml/_binary/cmvm/indexers.cc
#include "indexers.hh"
#include "bit_decompose.hh"
#include <cmath>
#include <algorithm>
#include <limits>

Pair idx_mc(const DAState &state) {
    Pair best_pair = {-1, -1, false, 0};
    size_t max_freq = 0;
    for (const auto &kv : state.freq_stat) {
        if (kv.second >= max_freq) {
            max_freq = kv.second;
            best_pair = kv.first;
        }
    }
    return best_pair;
}

Pair idx_mc_dc(const DAState &state, bool absolute) {
    Pair best_pair = {-1, -1, false, 0};
    float factor = 1e9; // Large factor to prioritize latency difference

    float max_score = absolute ? 0 : -std::numeric_limits<float>::infinity();
    for (const auto &kv : state.freq_stat) {
        auto &k = kv.first;
        float lat0 = state.ops[k.id0].latency;
        float lat1 = state.ops[k.id1].latency;
        float score = kv.second - factor * std::abs(lat0 - lat1);
        if (score >= max_score) {
            max_score = score;
            best_pair = k;
        }
    }
    return best_pair;
}

static inline int8_t qint_to_bw(const QInterval &q) {
    if (q.min == q.max && q.min == 0) {
        return 0;
    }
    int8_t keep_negative = q.min < 0 ? 1 : 0;
    float _min = q.min / q.step, _max = q.max / q.step;
    int8_t mbits = std::max(iceil_log2(_min), iceil_log2(_max + 1));
    return mbits + keep_negative;
}

std::tuple<int8_t, int8_t, int8_t>
overlap_counts(const QInterval &q0, const QInterval &q1, const int8_t shift1) {
    int8_t r0 = -get_lsb_loc(q0.step), r1 = -get_lsb_loc(q1.step) - shift1;
    int8_t b0 = qint_to_bw(q0), b1 = qint_to_bw(q1);
    int8_t l0 = r0 - b0, l1 = r1 - b1;
    std::array<int8_t, 4> sorted = {l0, r0, l1, r1};
    std::sort(sorted.begin(), sorted.end());
    if (r0 < l0 || r1 < l0) {
        std::swap(sorted[1], sorted[2]);
    }

    int8_t overlap = sorted[2] - sorted[1];
    int8_t accum = sorted[1] - sorted[0];
    int8_t tail = sorted[3] - sorted[2];
    return {accum, overlap, tail};
}

Pair idx_wmc(const DAState &state) {
    int64_t max_score = 0;
    Pair best_pair = {-1, -1, false, 0};
    for (const auto &kv : state.freq_stat) {
        auto &k = kv.first;
        auto [_accum, n_overlap, _tail] = overlap_counts(
            state.ops[k.id0].qint, state.ops[k.id1].qint, state.ops[k.id1].data
        );
        int64_t score = int64_t(kv.second) * n_overlap;
        if (score >= max_score) {
            max_score = score;
            best_pair = k;
        }
    }
    return best_pair;
}

Pair idx_wmc_dc(const DAState &state, bool absolute) {
    float max_score = absolute ? 0 : -std::numeric_limits<float>::infinity();
    Pair best_pair = {-1, -1, false, 0};
    for (const auto &kv : state.freq_stat) {
        auto &k = kv.first;
        auto [_accum, n_overlap, _tail] = overlap_counts(
            state.ops[k.id0].qint, state.ops[k.id1].qint, state.ops[k.id1].data
        );
        float lat0 = state.ops[k.id0].latency;
        float lat1 = state.ops[k.id1].latency;
        float score = kv.second * n_overlap - 256 * std::abs(lat0 - lat1);
        if (score >= max_score) {
            max_score = score;
            best_pair = k;
        }
    }
    return best_pair;
}
